### PR TITLE
[TRTLLM-11045][feat] Integrate SA with EAGLE3 and PARD

### PR DESCRIPTION
**Pull request by cascade812**
## Description

Integrate SA with EAGLE3 and PARD

Usage example 
```
eagle_model_dir = f"{llm_models_root()}/EAGLE3-LLaMA3.1-Instruct-8B"
target_model_dir = f"{llm_models_root()}/llama-3.1-model/Llama-3.1-8B-Instruct"

spec_config = Eagle3DecodingConfig(max_draft_len=4,
                                    speculative_model=eagle_model_dir,
                                    eagle3_one_model=True,
                                    use_sa_spec=True)
llm = LLM(model=target_model_dir,
            **pytorch_config,
            kv_cache_config=kv_cache_config,
            speculative_config=spec_config)            
```

### Configuration

| Setting | Value |
|---------|-------|
| **EAGLE3 Target/Draft** | Llama-3.1-8B-Instruct / EAGLE3-LLaMA3.1-Instruct-8B with draft length = 4 |
| **PARD Target/Draft** | Llama-3.1-8B-Instruct / PARD-Llama-3.2-1B with draft length = 4 |
| **SA Threshold** | 4 |
| **GPU** | single H100 |

### Dataset Details

| Metric | Value |
|--------|-------|
| **Dataset** | glaiveai/code_edits_sample |
| **Number of Requests** | 100 (randomly sampled) |
| **Average Input Length** | 390.9 tokens |
| **Average Output Length** | 256.0 tokens |

### Comparison Table

| Configuration | Total Output Throughput (tok/s) | TTFT (ms) | TPOT (ms) | Accept Rate | Accept Length |
|---------------|--------------------------------|-----------|-----------|-------------|---------------|
| EAGLE3 | 321.21 | 32.41 | 3.00 | 0.4726 | 2.89 |
| EAGLE3 + SA | **386.33** | 32.30 | **2.47** | **0.6459** | **3.58** |
| PARD | 343.63 | 35.85 | 2.78 | 0.5335 | 3.13 |
| PARD + SA | **415.48** | 36.13 | **2.27** | **0.7248** | **3.90** |
| Baseline | 150.33 | 25.00 | 6.58 | N/A | N/A |

### EAGLE3 + SA Improvement over EAGLE3
- **Total Output Throughput:** +20.3%
- **TPOT:** +17.6%
- **Acceptance Rate:** 64.59% vs 47.26% (+36.7%)
- **Acceptance Length:** 3.58 vs 2.89 tokens (+24.0%)

#### PARD + SA Improvement over PARD
- **Total Output Throughput:** +20.9%
- **TPOT:** +18.2% 
- **Acceptance Rate:** 72.48% vs 53.35% (+35.8%)
- **Acceptance Length:** 3.90 vs 3.13 tokens (+24.4%)

### Speedup vs Baseline
| Configuration | Speedup over Baseline (150.33 tok/s) |
|---------------|---------|
| EAGLE3 | **2.14x** |
| EAGLE3 + SA | **2.57x** |
| PARD | **2.29x** |
| PARD + SA | **2.76x** |

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

## Release Notes

* **New Features**
  * Suffix Automaton (SA) speculative decoding enhancement can now be combined with EAGLE3, PARD, and MTP decoders through new configuration flags (use_sa_spec, sa_spec_threshold)

* **Documentation**
  * Added comprehensive guidance on configuring and using SA speculative decoding across different decoding methods

* **Tests**
  * Added accuracy verification tests for SA speculative decoding with EAGLE3 and PARD decoders

### CODE REVIEW

**Review by ziyixiong-nv — APPROVED** [review]
LGTM

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 1]
```diff
@@ -311,6 +311,32 @@ def test_eagle3(self, overlap_scheduler, eagle3_one_model,
             task = GSM8K(self.MODEL_NAME)
             task.evaluate(llm)
 
+    @skip_pre_hopper
+    def test_eagle3_sa(self):
```
Comment (crazydemo) [inline tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Could you please add the new cases into qa/llm_function_core.txt?

**Review by Funatiq — APPROVED** [review]

**Review by pcastonguay — APPROVED** [review]

File: tensorrt_llm/_torch/speculative/pard.py [hunk 2]
```diff
@@ -193,6 +221,18 @@ def forward(
             )
             accepted_tokens = torch.cat([accepted_tokens, acc_padding], dim=1)
 
+        sa_manager = spec_metadata._get_sa_manager() if self.sa_enhancer else None
```
Comment (venkywonka) [inline tensorrt_llm/_torch/speculative/pard.py]: don't know anything about pard - but devin commented this - please check for relevance: 

PARD+SA passes wrongly-strided accepted_tokens to SA extend kernel
In the PARD+SA path, accepted_tokens is padded from shape (batch, K+1) to (batch, 2K) at pard.py:222 before being passed to sa_enhancer.extend_and_prepare at pard.py:226-234. The extend_and_prepare method forwards accepted_tokens[num_contexts:] (shape [num_gens, 2K]) to sa_manager.extend(), which passes the raw data_ptr to the CUDA kernel. The kernel at suffixAutomatonKernels.cu:61 indexes with acceptedTokensIn[i * (draftLength + 1) + j], assuming row stride K+1. However, the actual memory stride is 2K. For any K > 1 and batch index i > 0, the kernel reads from incorrect memory offsets, corrupting SA automaton states and producing wrong draft tokens. This could also cause out-of-bounds memory reads.

**Review by laikhtewari — APPROVED** [review]

**Review by hchings — APPROVED** [review]

**Review by StanleySun639 — APPROVED** [review]

**Review by venkywonka — APPROVED** [review]
